File: 0735-asteroid-collision/0735-asteroid-collision.cpp
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        vector<int> st;
        for (int a : asteroids) {
            while (!st.empty() && st.back() > 0 && a < 0) {
                if (st.back() < -a) {
                    st.pop_back(); // current left asteroid destroyed
                    continue;      // check again with new top
                } else if (st.back() == -a) {
                    st.pop_back(); // both destroy each other
                }
                a = 0; // current asteroid destroyed
                break;
            }
            if (a != 0) st.push_back(a); // survives
        }
        return st;  // ✅ return final result
    }
};

File: 0735-asteroid-collision/0735-asteroid-collision.java
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        Stack<Integer> stack = new Stack<>();
        for(int num:asteroids)
        {
            boolean destroyed = false;
            while(!stack.isEmpty() && num<0 && stack.peek()>0)
            {
                if(stack.peek()<-num)
                {
                    stack.pop();
                }
                else if(stack.peek()== -num)
                {
                    stack.pop();
                    destroyed=true;
                    break;
                }
                else
                {
                    destroyed = true;
                    break;
                }
            }
            if(!destroyed)
            {
                stack.push(num);
            }
        }
        int[] res = new int[stack.size()];
        for(int i=stack.size()-1;i>=0;i--)
        {
            res[i]=stack.pop();
        }
        return res;
    }
}